### PR TITLE
Geometry_Engine: Area for PlanarSurface and PolySurface

### LINKED ISSUE #1567: Geometry_Engine: Area for PlanarSurface
#### Description:
Implement a method to get the area from a PlanarSurface.
Should be as simple as taking the area of its curves.

Could preferably also add a fallback case for `IArea()` as it accepts any `IGeometry`

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR
Closes #1567 
As title,

Am open to handle the fallback in some other way if there is some desire for that.

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/EmZSNpJtCQFGvVK02yM_LYcB06hr3qx9b4AoKPeSbr4vYw?e=XLmfRZ

### Changelog

- and a private fallback for error messaging

### CODE REVIEW

File: Geometry_Engine/Query/Area.cs [hunk 1]
```diff
@@ -190,6 +207,16 @@ public static double IArea(this IGeometry geometry)
             return Area(geometry as dynamic);
         }
 
+
+        /***************************************************/
+        /**** Private Methods - Fallbacks               ****/
+        /***************************************************/
+
+        public static double Area(this IGeometry geometry)
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Geometry_Engine/Query/Area.cs]: done

File: Geometry_Engine/Query/Area.cs [hunk 2]
```diff
@@ -190,6 +207,16 @@ public static double IArea(this IGeometry geometry)
...
+        {
+            throw new NotImplementedException("Area for " + geometry.GetType().Name + " is not implemented.");
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Geometry_Engine/Query/Area.cs]: done

File: Geometry_Engine/Query/Area.cs [hunk 3]
```diff
@@ -190,6 +207,16 @@ public static double IArea(this IGeometry geometry)
             return Area(geometry as dynamic);
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Geometry_Engine/Query/Area.cs]: done

**Comment by pawelbaran** [comment]
> I think the `Area` method returns wrong results for `PlanarSurface` when `InternalBoundaries` overlap.
> Adding the test in seconds

Nice catch @epignatelli! However, I think this one can wait as it should be resolved by https://github.com/BHoM/BHoM/issues/676

**Comment by adecler** [comment]
@FraserGreenroyd , I am not sure how BHoM/BHoM#676 is going to address the comment from @epignatelli. I don't feel I should approve this until Eduardo had a chance to respond.

**Comment by kThorsager** [comment]
> @FraserGreenroyd , I am not sure how [BHoM/BHoM#676](https://github.com/BHoM/BHoM/issues/676) is going to address the comment from @epignatelli. I don't feel I should approve this until Eduardo had a chance to respond.

If it is Immutable we would ensure its validity at creation and not need to check for overlapping internal boundaries for each operation which might be affected

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd , I am not sure how [BHoM/BHoM#676](https://github.com/BHoM/BHoM/issues/676) is going to address the comment from @epignatelli. I don't feel I should approve this until Eduardo had a chance to respond.

That's fair enough @adecler - I only requested your review as @pawelbaran said he was leaving a final vote to you and @al-fisher on a decision so figured it was the best way for you to see and case your vote as appropriate :smile:

**Comment by epignatelli** [comment]
Thanks @adecler! I think that, when`PlanarSurface` becomes `IImmutable`, and at construction time we make sure that we solve any edge cases for both internal and external boundaries, this should be solved. 

My suggestion for this pr would be to merge, and raise an issue about the bug above, reference and solve it together with BHoM/BHoM#676, maybe sooner rather than later.

WDYT?

**Comment by adecler** [comment]
@epignatelli, If you're happy, I'm fine too. 

It wasn't just made clear in the reference issue that it would also take care of the overlapping boundaries so wanted to highlight that.

**Comment by kThorsager** [comment]
Updated https://github.com/BHoM/BHoM/issues/676 with a comment clarifying what the scope should be

File: Geometry_Engine/Query/Area.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/Area.cs]: do not want to keep the discussion regarding this open forever, but every single other method that I have looked at in Geometry_Engine has the interface method at the bottom. The same has been done in structure_Engine (did that to align with the initial setup in geometry et al).

Do we have any place where the opposite convention has been used?

**Pull request merged**